### PR TITLE
DEPR: __setitem__ on dt64tz with mixed timezones

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/_libs/tslib.pyx [hunk 1]
```diff
@@ -862,3 +863,50 @@ cdef inline bint _parse_today_now(str val, int64_t* iresult, bint utc):
...
+        if checknull_with_nat_and_na(item):
+            # this catches pd.NA which would raise in the Timestamp constructor
+            ival = NPY_NAT
+
+        else:
+            ts = Timestamp(item)
+            if ts is NaT:
+                ival = NPY_NAT
+            else:
+                if ts.tz is not None:
```
Comment (mroeschke) [inline pandas/_libs/tslib.pyx:899]: Could we shortcut & skip if `ts.tz == tz`?

File: pandas/tests/series/methods/test_astype.py [hunk 2]
```diff
@@ -97,6 +97,20 @@ def test_astype_dict_like(self, dtype_class):
...
 
 class TestAstype:
+    def test_astype_mixed_object_to_dt64tz(self):
+        # pre-2.0 this raised ValueError bc of tz mismatch
+        # xref GH#32581
+        ts = Timestamp("2016-01-04 05:06:07", tz="US/Pacific")
+        ts2 = ts.tz_convert("Asia/Tokyo")
+
+        ser = Series([ts, ts2], dtype=object)
+        res = ser.astype("datetime64[ns, Europe/Brussels]")
```
Comment (mroeschke) [inline pandas/tests/series/methods/test_astype.py:107]: Whats the error message if we try this when an element is not convertable e.g. `ser = Series([ts, object], dtype=object)`

Comment (jbrockmendel) [reply pandas/tests/series/methods/test_astype.py:107]: `ValueError: could not convert string to Timestamp`

File: pandas/_libs/tslib.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx:899]: probably but i think if/when we get around to optimizing this we'll skip Timestamp altogether

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**